tests/apps/actorfeatures/app.go

```diff
@@ -0,0 +1,345 @@
+var actors = map[string]daprActor{}
+var actorsMutex = &sync.Mutex{}
+
+func resetLogs() {
+	logsMutex.Lock()
+	defer logsMutex.Unlock()
+
+	logs = []actorLogEntry{}
+}
+
+func appendLog(actorType string, actorID string, action string, start int) {
+	logEntry := actorLogEntry{
+		Action:         action,
+		ActorType:      actorType,
+		ActorID:        actorID,
+		StartTimestamp: start,
+		EndTimestamp:   epoch(),
+	}
+
+	logsMutex.Lock()
+	defer logsMutex.Unlock()
+	logs = append(logs, logEntry)
+}
+
+func getLogs() []actorLogEntry {
+	return logs
+}
+
+func createActorID(actorType string, id string) string {
+	return fmt.Sprintf("%s.%s", actorType, id)
+}
+
+// indexHandler is the handler for root path
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	log.Println("indexHandler is called")
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func logsHandler(w http.ResponseWriter, r *http.Request) {
+	log.Printf("Processing dapr %s request for %s", r.Method, r.URL.RequestURI())
+	if r.Method == "DELETE" {
+		resetLogs()
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(getLogs())
+}
+
+func configHandler(w http.ResponseWriter, r *http.Request) {
+	log.Printf("Processing dapr request for %s", r.URL.RequestURI())
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(daprConfigResponse)
+}
+
+func actorMethodHandler(w http.ResponseWriter, r *http.Request) {
+	log.Printf("Processing actor method request for %s", r.URL.RequestURI())
+
+	start := epoch()
+
+	actorType := mux.Vars(r)["actorType"]
+	id := mux.Vars(r)["id"]
+	method := mux.Vars(r)["method"]
+	reminderOrTimer := mux.Vars(r)["reminderOrTimer"] != ""
+
+	hostname, err := os.Hostname()
+	var data []byte
+	if method == "hostname" {
+		data = []byte(hostname)
+	} else {
+		// Sleep for all calls, except timer and reminder.
+		if !reminderOrTimer {
+			time.Sleep(secondsToWaitInMethod * time.Second)
+		}
+		data, err = json.Marshal(response{
+			actorType,
+			id,
+			method,
+			start,
+			epoch(),
+			"",
+		})
+	}
+
+	if err != nil {
+		fmt.Printf("Error: %v", err.Error())
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	appendLog(actorType, id, method, start)
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(daprActorResponse{
+		Data: data,
+	})
+}
+
+func activateDeactivateActorHandler(w http.ResponseWriter, r *http.Request) {
+	log.Printf("Processing %s actor request for %s", r.Method, r.URL.RequestURI())
+
+	start := epoch()
+
+	actorType := mux.Vars(r)["actorType"]
+	id := mux.Vars(r)["id"]
+
+	if actorType != registedActorType {
+		log.Printf("Unknown actor type: %s", actorType)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	actorID := createActorID(actorType, id)
+
+	actorsMutex.Lock()
+	defer actorsMutex.Unlock()
```

Comment:
Do we need mutex here? I think dapr runtime must handle deactivation and activation order properly. cc/ @yaron2 @amanbha

Reply:
So, since I am exposing a HTTP endpoint, it is a good practice to handle concurrent calls correctly. Even if the client itself is behaving as expected. There is a similar discussion about this on the java-sdk. Relying on the http client's behavior to handle concurrency is not enough IMO - what if the message is retried? If a code is handling a shared state between threads, it should be thread safe.

Reply:
There is another point: does the runtime make sure only one activation per actor type? If it enforces per actor-actorId instead, then this mutex is need to handle that.

Again, this mutex makes the code thread-safe, which does not hurt since there is no IO in this code.

Reply:
Just to be clear, I am locking here because the map is shared for all actors and I am modifying the map here. This is a reference I found on thread-safety for map in Go:
https://github.com/orcaman/concurrent-map

Reply:
> There is another point: does the runtime make sure only one activation per actor type? If it enforces per actor-actorId instead, then this mutex is need to handle that.
> 
> Again, this mutex makes the code thread-safe, which does not hurt since there is no IO in this code.

The Dapr runtime handles concurrency and single activations for `actorType-actorID`. Meaning, only a single instance of an actor ID for a given actor Type will be activated anywhere in the cluster.

Reply:
> Just to be clear, I am locking here because the map is shared for all actors and I am modifying the map here. This is a reference I found on thread-safety for map in Go:
> https://github.com/orcaman/concurrent-map

Better to use the built on [sync.Map](https://golang.org/pkg/sync/#Map)

Reply:
I can do it if I use append only: see the reference above. It has a comment on that: Prior to Go 1.9, there was no concurrent map implementation in the stdlib. In Go 1.9, sync.Map was introduced. The new sync.Map has a few key differences from this map. The stdlib sync.Map is designed for append-only scenarios. So if you want to use the map for something more like in-memory db, you might benefit from using our version. You can read more about it in the golang repo, for example here and here

Reply:
Look at sync.Map. It has a Store and LoadOrStore methods. Store will overwrite a value for a given key, and LoadOrStore will do the same.

Reply:
Looking at that now. I want to use it, so I can push the lock low in the stack and have this code thread-safe without locking per actor id.

Reply:
Dapr runtime, will make only 1 concurrent call per actor id.

Reply:
Using sync.Map now. Let me run e2e tests to make sure it works.